### DISCUSSION TITLE
Pong ball collisions

### DISCUSSION THREAD

**Discussion by TFDxSnAKe** [MonoGame/MonoGame#8488] [General]
I am fairly new to coding and I'm currently working on a pong clone in monogame. The first thing I tried working on is the ball itself.

I'm probably doing something wrong but I can't get the ball to bounce whenever it reaches the screen borders, I want it to flip either the X or Y value, depending on which side it bounces on, by multiplying it by -1 but the ball just keep sticking at the borders.
This is the code that I have:

using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using Microsoft.Xna.Framework.Input;

namespace ball_Test
{
    public class Game1 : Game
    {
        private GraphicsDeviceManager graphics;
        private SpriteBatch spriteBatch;
        private Texture2D Pong_ball;
        Vector2 middle;
        Vector2 flip_y;
        Vector2 flip_x;

        private int speed = 200;
        private int deltaspeed;
        private int ball_x;
        private int ball_y;
        private int ball_dx;
        private int ball_dy;

        public Game1()
        {
            graphics = new GraphicsDeviceManager(this);
            Content.RootDirectory = "Content";
            IsMouseVisible = true;
        }

        protected override void Initialize()
        {
            // TODO: Add your initialization logic here

            base.Initialize();
        }

        protected override void LoadContent()
        {
            spriteBatch = new SpriteBatch(GraphicsDevice);
            Pong_ball = Content.Load<Texture2D>("pong_ball");
            middle = new Vector2(graphics.PreferredBackBufferWidth / 2, graphics.PreferredBackBufferHeight / 2);
            flip_x = new Vector2(-1, 1);
            flip_y = new Vector2(1, -1);

            // TODO: use this.Content to load your game content here
        }

        protected override void Update(GameTime gameTime)
        {
            ball_dx = 1;
            ball_dy = 1;

            deltaspeed = (int)(speed * (float)gameTime.ElapsedGameTime.TotalSeconds);
            middle.X += ball_dx * deltaspeed;
            middle.Y += ball_dy * deltaspeed;

            if (middle.Y > graphics.PreferredBackBufferHeight - Pong_ball.Height || middle.Y < 0)
            {
                //middle *= flip_y;
                middle.Y *= -1;
                //ball_dy = -ball_dy;
            }

            if (middle.X > graphics.PreferredBackBufferWidth - Pong_ball.Width || middle.X < 0)
            {
                middle.X *= -1;
            }

            base.Update(gameTime);
        }

        protected override void Draw(GameTime gameTime)
        {
            GraphicsDevice.Clear(Color.Gray);

            spriteBatch.Begin();
            spriteBatch.Draw(Pong_ball, middle, Color.White);
            spriteBatch.End();

            base.Draw(gameTime);
        }
    }
}

What am I doing wrong???

**Comment by KelsamGames** [comment]
To solve your code issues specifically; you have a couple problems.
1. You need to reverse your velocity (ball_dx/ball_dy), not your position (middle), when collision is detected.
2. You can't set your velocity (ball_dx/ball_dy) at the start of every frame or your ball with never change direction.
3. You need to mod the position to "correct" the collision so it isn't detected multiple frames in a row (middle.X = graphics.PreferredBackBufferWidth - Pong_ball.Width, when detected collision to the right for example).

Also, your "middle" variable isn't using an offset, so it is actually the top-left of the ball. Probably rename that to something like "ballPosition" so it is less confusing. You have variables like "ball_x" and "ball_y" that are never used. That, along with the issues named above, tells me that your key issue here is probably that you don't fully understand what your variables are representing and/or how they're used, and instead are just trying to "twist every nob" until you get the desired result. So, I'd recommend trying to address that. Play around with these variables and read the tooltips from VS on the MG functions you're calling to really understand what they're doing. Think about the game in a frame-to-frame context and how the variables change each frame. For example, if in one frame your ball moves into a collision space by 10 and you don't correct the position and only set your speed/velocity vars to reverse (by multiplying by -1), then the next frame your speed is only 5 so after moving the ball it is still in the collision space and will trigger again and then erroneously revert your speed again sending it further into the collision space.
